examples/multigpu_advtrain/run.py

```diff
@@ -0,0 +1,317 @@
+"""
+This example adversarially trains a model using iterative attacks.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import six
+import math
+import time
+import os
+import logging
+import random
+
+import tensorflow as tf
+from tensorflow.python.platform import app
+from tensorflow.python.platform import flags
+
+from cleverhans.utils_mnist import data_mnist
+from cleverhans.utils_tf import model_loss
+from cleverhans.utils_tf import batch_indices
+from cleverhans.utils_tf import _FlagsWrapper
+import cifar_input
+
+from make_model import make_model
+from evaluator import Evaluator, get_attack
+
+
+FLAGS = flags.FLAGS
+_data_path = {'svhn':  '/ssd1/datasets/svhn/',
+              'cifar10':  '/ssd1/datasets/cifar-10/'}
+if "CIFAR10_PARDIR" in os.environ:
+    _data_path['cifar10'] = os.environ['CIFAR10_PARDIR']
+
+
+def preprocess_batch(images_batch, preproc_func=None):
+    if preproc_func is None:
+        return images_batch
+
+    with tf.variable_scope('preprocess'):
+        images_list = tf.split(images_batch, int(images_batch.shape[0]))
+        result_list = []
+        for img in images_list:
+            reshaped_img = tf.reshape(img, img.shape[1:])
+            processed_img = preproc_func(reshaped_img)
+            result_list.append(tf.expand_dims(processed_img, axis=0))
+        result_images = tf.concat(result_list, axis=0)
+    return result_images
+
+
+def model_train(sess, model, x_pre, x, y, predictions, X_train, Y_train,
+                predictions_adv=None, init_all=True, evaluate=None,
+                verbose=True, args=None, hparams=None):
```

Comment:
I imagine this one really does have to be copy-pasted since the multi-GPU training logic is probably quite a lot different from the existing `model_train`, but it would be good for the docstring to mention specifically how this one is different.

Reply:
This is now part of TrainManager and TrainerMultiGPU and TrainerSingleGPU with most of the code being shared. The refactoring is done in commits between 4bd545ec41930128bfbc397fd34699e3c9e2c0e2 and f75fc0d7d4c361685578685f302a522bbd4fb2fa.